onnxruntime/python/onnxruntime_pybind_state.cc

```diff
@@ -2907,16 +2941,44 @@ class EnvInitializer {
     return env_holder.Get();
   }
 
+  static bool GetUsePerSessionThreads() {
+    return use_per_session_threads;
+  }
+
  private:
   EnvInitializer() {
     std::unique_ptr<Environment> env_ptr;
     Env::Default().GetTelemetryProvider().SetLanguageProjection(OrtLanguageProjection::ORT_PROJECTION_PYTHON);
-    OrtPybindThrowIfError(Environment::Create(std::make_unique<LoggingManager>(
-                                                  std::make_unique<CLogSink>(),
-                                                  Severity::kWARNING, false, LoggingManager::InstanceType::Default,
-                                                  &SessionObjectInitializer::default_logger_id),
-                                              env_ptr));
+
+    // create logging manager here
+    std::unique_ptr<LoggingManager> lm = std::make_unique<LoggingManager>(
+        std::make_unique<CLogSink>(),
+        Severity::kWARNING, false, LoggingManager::InstanceType::Default,
+        &SessionObjectInitializer::default_logger_id);
+
+    if (EnvInitializer::use_per_session_threads) {
+      OrtPybindThrowIfError(Environment::Create(std::move(lm),
+                                                env_ptr));
+    } else {
+      OrtPybindThrowIfError(Environment::Create(std::move(lm),
+                                                env_ptr,
+                                                &EnvInitializer::tp_options,
+                                                true));
+    }
+
     session_env_ = std::shared_ptr<Environment>(env_ptr.release());
+
+    if (!EnvInitializer::use_per_session_threads) {
+      py::object atexit_register = py::module::import("atexit").attr("register");
```

Comment:
I've never used pybind much, I'm assuming this works because we'd only ever execute this code from within the context of a running Python interpreter that called `set_global_thread_pool_size` at some point?

Reply:
atexit would not work good in this case. Because, you don't know if there is any ONNX Runtime python objects around when the atexit callback is executing.

Reply:
Maybe we downgrade the shared_ptr stored in EnvInitializer to a weak_ptr? Then every time the lock comes back null, we run the actual initialization code. Would also ensure the pool gets cleaned up when all the inference sessions get deleted.